crates/engine/tree/src/tree/payload_processor/multiproof.rs

```diff
@@ -997,147 +1017,187 @@ impl MultiProofTask {
-                            // record the wait time
-                            self.metrics
-                                .first_update_wait_time_histogram
-                                .record(start.elapsed().as_secs_f64());
-                            first_update_time = Some(Instant::now());
-                            debug!(target: "engine::tree::payload_processor::multiproof", "Started state root calculation");
-                        }
 
-                        let account_targets = targets.len();
-                        let storage_targets =
-                            targets.values().map(|slots| slots.len()).sum::<usize>();
-                        prefetch_proofs_requested += self.on_prefetch_proof(targets);
-                        trace!(
-                            target: "engine::tree::payload_processor::multiproof",
-                            account_targets,
-                            storage_targets,
-                            prefetch_proofs_requested,
-                            "Prefetching proofs"
-                        );
-                    }
-                    MultiProofMessage::StateUpdate(source, update) => {
-                        trace!(target: "engine::tree::payload_processor::multiproof", "processing MultiProofMessage::StateUpdate");
-                        if first_update_time.is_none() {
-                            // record the wait time
-                            self.metrics
-                                .first_update_wait_time_histogram
-                                .record(start.elapsed().as_secs_f64());
-                            first_update_time = Some(Instant::now());
-                            debug!(target: "engine::tree::payload_processor::multiproof", "Started state root calculation");
-                        }
-
-                        let len = update.len();
-                        state_update_proofs_requested += self.on_state_update(source, update);
-                        trace!(
-                            target: "engine::tree::payload_processor::multiproof",
-                            ?source,
-                            len,
-                            ?state_update_proofs_requested,
-                            "Received new state update"
-                        );
-                    }
-                    MultiProofMessage::FinishedStateUpdates => {
-                        trace!(target: "engine::tree::payload_processor::multiproof", "processing MultiProofMessage::FinishedStateUpdates");
-                        updates_finished = true;
-                        updates_finished_time = Some(Instant::now());
-                        if self.is_done(
-                            proofs_processed,
-                            state_update_proofs_requested,
-                            prefetch_proofs_requested,
-                            updates_finished,
-                        ) {
+            crossbeam_channel::select! {
```

Comment:
I think it's a very nice change: we now have a channel that we receive the state updates on, and a separate channel that we receive the proof results on. Much cleaner imo, we can maybe separate the handling of messages into separate methods so it's easier to navigate, but not a blocker

Reply:
this makes sense, let me configure and see how to clean that up 

> we can maybe separate the handling of messages into separate methods so it's easier to navigate, but not a blocker

Reply:
@shekhirin just did a prototype, i think its easier to review in a separate pr :) noted down in https://github.com/paradigmxyz/reth/issues/19182